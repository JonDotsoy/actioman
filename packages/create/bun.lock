{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "create-actioman",
      "dependencies": {
        "@jondotsoy/flags": "^2.3.0",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "^5.0.0",
      },
    },
  },
  "packages": {
    "@jondotsoy/console-draw": ["@jondotsoy/console-draw@0.1.2", "", { "dependencies": { "@jondotsoy/style-text": "^0.9.1" }, "peerDependencies": { "prettier": "^3.3.3", "typescript": "^5.0.0" } }, "sha512-BBZGLd6RzjKxKcgD2OwD7v6GeKlf53Ysk59H+7+HvyGLcJZ7AJk8rOCq/yv813KbWV2teTCEBQXnnP5hZe/2CA=="],

    "@jondotsoy/flags": ["@jondotsoy/flags@2.3.0", "", { "dependencies": { "@jondotsoy/console-draw": "^0.1.2" } }, "sha512-vUMe4NdXKv8RxMXrKCJLsHBt3XQ8/YV4Fyr1I43vSLe5v0M8sdY3wJl1qdTgwRBZES2zMLoenxOUSNGE9r/5NA=="],

    "@jondotsoy/style-text": ["@jondotsoy/style-text@0.9.2", "", { "peerDependencies": { "typescript": "^5.0.0" } }, "sha512-6fonOzrANwrw+UIM4fivtDZLvuCN00yH0lCEYi14ln4j8fSHZokb6TLrqOu38GCqqQnZ2LuzJTQBWgv7jN6iVA=="],

    "@types/bun": ["@types/bun@1.2.2", "", { "dependencies": { "bun-types": "1.2.2" } }, "sha512-tr74gdku+AEDN5ergNiBnplr7hpDp3V1h7fqI2GcR/rsUaM39jpSeKH0TFibRvU0KwniRx5POgaYnaXbk0hU+w=="],

    "@types/node": ["@types/node@22.13.1", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-jK8uzQlrvXqEU91UxiK5J7pKHyzgnI1Qnl0QDHIgVGuolJhRb9EEl28Cj9b3rGR8B2lhFCtvIm5os8lFnO/1Ew=="],

    "@types/ws": ["@types/ws@8.5.14", "", { "dependencies": { "@types/node": "*" } }, "sha512-bd/YFLW+URhBzMXurx7lWByOu+xzU9+kb3RboOteXYDfW+tr+JZa99OyNmPINEGB/ahzKrEuc8rcv4gnpJmxTw=="],

    "bun-types": ["bun-types@1.2.2", "", { "dependencies": { "@types/node": "*", "@types/ws": "~8.5.10" } }, "sha512-RCbMH5elr9gjgDGDhkTTugA21XtJAy/9jkKe/G3WR2q17VPGhcquf9Sir6uay9iW+7P/BV0CAHA1XlHXMAVKHg=="],

    "prettier": ["prettier@3.4.2", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-e9MewbtFo+Fevyuxn/4rrcDAaq0IYxPGLvObpQjiZBMAzB9IGmzlnG9RZy3FFas+eBMu2vA0CszMeduow5dIuQ=="],

    "typescript": ["typescript@5.7.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-84MVSjMEHP+FQRPy3pX9sTVV/INIex71s9TL2Gm5FG/WG1SqXeKyZ0k7/blY/4FdOzI12CBy1vGc4og/eus0fw=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],
  }
}
