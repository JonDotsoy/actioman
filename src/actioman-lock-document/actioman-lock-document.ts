import type { JsonDBRecord } from "./dtos/json-db-record";

export class ActionmanLockDocument {
  constructor(private recordsList: JsonDBRecord[] = []) {}
  keys() {
    return this.recordsList.map((e) => e.key);
  }
  set(key: string[], value: any) {
    const v: null | JsonDBRecord = this.get(key);
    if (v) v.value = value;
    else this.recordsList.push({ key, value });
  }
  get(key: string[]) {
    return (
      this.recordsList.find(
        (e) => JSON.stringify(e.key) === JSON.stringify(key),
      ) ?? null
    );
  }
  has(key: string[]) {
    return this.get(key) !== null;
  }
  delete(key: string[]) {
    this.recordsList = this.recordsList.filter(
      (e) => JSON.stringify(e.key) !== JSON.stringify(key),
    );
  }
  toJSON() {
    type NestedRecord = Record<string, any>;
    const keyRefs = new Map<string, string[]>();
    const nestedRefs = new WeakMap<string[], NestedRecord>();
    const staticKey = (key: string[]): string[] => {
      const ref = JSON.stringify(key);
      if (keyRefs.has(ref)) return keyRefs.get(ref)!;
      keyRefs.set(ref, key);
      return key;
    };
    const getDeep = (key: string[]): NestedRecord => {
      const ref = staticKey(key);
      if (nestedRefs.has(ref)) return nestedRefs.get(ref)!;
      const newNestedRecord: NestedRecord = {};
      nestedRefs.set(ref, newNestedRecord);
      const parentKey = key.slice(0, -1);
      const pathKey = key.at(-1);
      if (pathKey) {
        getDeep(key.slice(0, -1))[pathKey] = newNestedRecord;
      }
      return newNestedRecord;
    };
    this.recordsList.forEach(({ key, value }) => {
      getDeep(key).toJSON = () => value;
    });
    return JSON.stringify(nestedRefs.get(staticKey([])), null, 2);
  }
  toJSONL() {
    return this.recordsList.map((e) => JSON.stringify(e)).join("\n");
  }
  toString() {
    return this.toJSONL();
  }
  static fromJSONL(payload: string) {
    const records = payload.split("\n").reduce((acc, e) => {
      try {
        return [...acc, JSON.parse(e)];
      } catch {}
      return acc;
    }, [] as JsonDBRecord[]);

    return new ActionmanLockDocument(records);
  }
  static fromJSON(payload: string) {
    const obj = JSON.parse(payload);
    const flat = function* (
      value: unknown,
      path: string[],
    ): Generator<JsonDBRecord> {
      if (typeof value === "object" && value !== null) {
        for (const [key, val] of Object.entries(value)) {
          yield* flat(val, [...path, key]);
        }
        return;
      }
      yield {
        key: path,
        value: value,
      };
    };

    return new ActionmanLockDocument(Array.from(flat(obj, [])));
  }
}
