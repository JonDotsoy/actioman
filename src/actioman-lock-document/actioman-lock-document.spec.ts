import { describe, it, expect } from "bun:test";
import { ActionmanLockDocument } from "./actioman-lock-document";

describe("ActionmanLockDocument", () => {
  it("toJSONL", () => {
    const actionmanLockDocument = new ActionmanLockDocument();

    actionmanLockDocument.set(["wrapp", "foo", "ok"], true);
    actionmanLockDocument.set(
      ["wrapp", "foo", "age"],
      "Thu, 23 Jan 2025 18:16:04 GMT",
    );
    actionmanLockDocument.set(["wrapps", "foo", "ok"], true);

    expect(actionmanLockDocument.toJSONL()).toMatchSnapshot();
  });

  it("fromJSONL", () => {
    const actionmanLockDocument = new ActionmanLockDocument();
    actionmanLockDocument.set(["wrapp", "foo", "ok"], true);
    actionmanLockDocument.set(
      ["wrapp", "foo", "age"],
      "Thu, 23 Jan 2025 18:16:04 GMT",
    );
    actionmanLockDocument.set(["wrapps", "foo", "ok"], true);
    const payload = actionmanLockDocument.toJSONL();

    expect(ActionmanLockDocument.fromJSONL(payload)).toMatchSnapshot();
  });

  it("toJSON", () => {
    const actionmanLockDocument = new ActionmanLockDocument();

    actionmanLockDocument.set(["wrapp", "foo", "ok"], true);
    actionmanLockDocument.set(
      ["wrapp", "foo", "age"],
      "Thu, 23 Jan 2025 18:16:04 GMT",
    );
    actionmanLockDocument.set(["wrapps", "foo", "ok"], true);

    expect(
      JSON.stringify(actionmanLockDocument.toJSON(), null, 2),
    ).toMatchSnapshot();
  });

  it("fromJSON", () => {
    const actionmanLockDocument = new ActionmanLockDocument();
    actionmanLockDocument.set(["wrapp", "foo", "ok"], 4);
    actionmanLockDocument.set(
      ["wrapp", "foo", "age"],
      "Thu, 23 Jan 2025 18:16:04 GMT",
    );
    actionmanLockDocument.set(["wrapps", "foo", "ok"], true);
    const payload = JSON.stringify(actionmanLockDocument.toJSON(), null, 2);

    expect(ActionmanLockDocument.fromJSON(payload)).toMatchSnapshot();
  });
});
