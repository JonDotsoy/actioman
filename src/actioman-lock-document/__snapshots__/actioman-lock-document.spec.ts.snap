// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionmanLockDocument toJSONL 1`] = `
"{"key":["wrapp","foo","ok"],"value":true}
{"key":["wrapp","foo","age"],"value":"Thu, 23 Jan 2025 18:16:04 GMT"}
{"key":["wrapps","foo","ok"],"value":true}"
`;

exports[`ActionmanLockDocument fromJSONL 1`] = `
ActionmanLockDocument {
  "recordsList": [
    {
      "key": [
        "wrapp",
        "foo",
        "ok",
      ],
      "value": true,
    },
    {
      "key": [
        "wrapp",
        "foo",
        "age",
      ],
      "value": "Thu, 23 Jan 2025 18:16:04 GMT",
    },
    {
      "key": [
        "wrapps",
        "foo",
        "ok",
      ],
      "value": true,
    },
  ],
}
`;

exports[`ActionmanLockDocument toJSON 1`] = `""{\\n  \\"wrapp\\": {\\n    \\"foo\\": {\\n      \\"ok\\": true,\\n      \\"age\\": \\"Thu, 23 Jan 2025 18:16:04 GMT\\"\\n    }\\n  },\\n  \\"wrapps\\": {\\n    \\"foo\\": {\\n      \\"ok\\": true\\n    }\\n  }\\n}""`;

exports[`ActionmanLockDocument fromJSON 1`] = `
ActionmanLockDocument {
  "recordsList": [
    {
      "key": [],
      "value": 
"{
  "wrapp": {
    "foo": {
      "ok": 4,
      "age": "Thu, 23 Jan 2025 18:16:04 GMT"
    }
  },
  "wrapps": {
    "foo": {
      "ok": true
    }
  }
}"
,
    },
  ],
}
`;
